Sources/Valkey/Subscriptions/SubscriptionConnectionStateMachine.swift

```diff
@@ -0,0 +1,258 @@
+            switch stateMachine.release(id: id) {
+            case .release(let continuation):
+                continuation.resume()
+                self.logger.trace("Released connection for subscriptions")
+            case .doNothing:
+                break
+            }
+        }
+    }
+
+    func cancelSubscriptionConnection(id: Int) {
+        self.logger.trace("Cancel subscription connection", metadata: ["valkey_subscription_connection_id": .stringConvertible(id)])
+        self.subscriptionConnectionStateMachine.withLock { stateMachine in
+            switch stateMachine.cancel(id: id) {
+            case .cancel(let cont):
+                cont.resume(throwing: CancellationError())
+            case .release(let continuation):
+                continuation.resume()
+                self.logger.trace("Released connection for subscriptions")
+            case .doNothing:
+                break
+            }
+        }
+    }
+}
+
+/// StateMachine for acquiring Subscription Connection.
+@usableFromInline
+struct SubscriptionConnectionStateMachine<Value, Request, ReleaseRequest>: ~Copyable {
+    enum State: ~Copyable {
+        /// We have no connection
+        case uninitialized
+        /// We are acquiring a connection
+        case acquiring([Int: Request])
+        /// We have a connection
+        case acquired(Value, Set<Int>, ReleaseRequest)
+    }
+    var state: State
+
+    init() {
+        self.state = .uninitialized
+    }
+
+    init(state: consuming State) {
+        self.state = state
+    }
+
+    enum GetAction {
+        case startAcquire
+        case doNothing
+        case completeRequest(Value)
+    }
+
+    mutating func get(id: Int, request: Request) -> GetAction {
+        switch consume self.state {
+        case .uninitialized:
+            self = .acquiring([id: request])
+            return .startAcquire
+        case .acquiring(var map):
+            map[id] = request
+            self = .acquiring(map)
+            return .doNothing
+        case .acquired(let connection, var ids, let releaseRequest):
+            ids.insert(id)
+            self = .acquired(connection, ids, releaseRequest)
+            return .completeRequest(connection)
+        }
+    }
+
+    enum CancelAction {
+        case cancel(Request)
+        case release(ReleaseRequest)
+        case doNothing
+    }
+
+    mutating func cancel(id: Int) -> CancelAction {
+        switch consume self.state {
+        case .uninitialized:
+            self = .uninitialized
+            return .doNothing
+        case .acquiring(var map):
+            guard let continuation = map.removeValue(forKey: id) else {
+                self = .acquiring(map)
+                return .doNothing
+            }
+            if map.isEmpty {
+                self = .uninitialized
+            } else {
+                self = .acquiring(map)
+            }
+            return .cancel(continuation)
```

Comment:
We should also cancel leasing the connection.  `pool.lease()` is cancellable.

Reply:
Not sure how we cancel the lease, it is running in the background discarding task group

Reply:
Ooh lovely